scio-core/src/main/scala/com/spotify/scio/values/SCollection.scala

```diff
@@ -920,6 +922,68 @@ sealed trait SCollection[T] extends PCollectionWrapper[T] {
     _.pApply(Sample.fixedSizeGlobally(sampleSize)).map(_.asScala)
   }
 
+  final private class WeightedHeapAggregator(totalWeight: Long, cost: T => Long)(implicit
```

Comment:
This is potentially a good idea, but for #5321 we explicitly want random elements, not the largest ones. i.e. we'd want `Aggregator[T, (Long, List[T]), Iterable[T]]` where the `Long` is just the total byte count added so far, where any additional elements over the target would just be dropped. 

For the Zstd use-case getting close to the target is good enough (thus the random sampling) but you could argue that for an aggregator you should keep trying to add elements until the List is as full as possible. :shrug:

Reply:
> where the Long is just the total byte count added so far

The `Long` in the priority queue is the result of `Random.nextLong()`, not the weight. This is the same strategy used for the `sample(sampleSize: Int)`  to select elements uniformly at random.

Reply:
Ahhh indeed, sorry, missed this